plugins/clickhouse/src/easy_client.cpp

```diff
@@ -0,0 +1,245 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2025 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+#include "easy_client.hpp"
+
+#include "clickhouse/client.h"
+#include "tenzir/detail/enumerate.hpp"
+#include "tenzir/view3.hpp"
+
+#include <boost/regex.hpp>
+
+#include <ranges>
+
+using namespace clickhouse;
+using namespace std::string_view_literals;
+
+/// The implementation works via multiple customization points:
+/// * Implementing the entire transformer by hand. This is done for Tuple and
+///   Array
+/// * A common implementation for non-structural types that works via
+///   * A trait to handle names and allocations.
+///     * This trait is auto-implemented for most types via a X-macro.
+///     * It is specialized for some types with special requirements
+///   * A `value_transform` function that translates between a tenzir::data
+///     value and the expected clickhouse API value.
+
+namespace tenzir::plugins::clickhouse {
+
+auto Easy_Client::make(Arguments args,
+                       diagnostic_handler& dh) -> std::unique_ptr<Easy_Client> {
+  auto client = std::make_unique<Easy_Client>(std::move(args), dh);
+  /// Note that technically, we have a ToCToU bug here. The table could be
+  /// created or deleted in between this, the `get` call below and the potential
+  /// creation in `insert`
+  const auto table_existed = client->table_exists();
+  TENZIR_TRACE("table exists: {}", table_existed);
+  if (client->args_.mode.inner == mode::create and table_existed) {
+    diagnostic::error("mode is `create`, but table `{}` already exists",
+                      client->args_.table.inner)
+      .primary(client->args_.mode)
+      .primary(client->args_.table)
+      .emit(client->dh_);
+    return nullptr;
+  }
+  if (client->args_.mode.inner == mode::create_append and not table_existed
+      and not client->args_.primary) {
+    diagnostic::error("table `{}` does not exist, but no `primary` was "
+                      "specified",
+                      client->args_.table.inner)
+      .primary(client->args_.table)
+      .emit(client->dh_);
+    return nullptr;
+  }
+  if (client->args_.mode.inner == mode::append and not table_existed) {
+    diagnostic::error("mode is `append`, but table `{}` does not exist",
+                      client->args_.table.inner)
+      .primary(client->args_.mode)
+      .primary(client->args_.table)
+      .emit(client->dh_);
+    return nullptr;
+  }
+  if (table_existed) {
+    if (not client->get_schema_transformations()) {
+      return nullptr;
+    }
+  }
+  return client;
+}
+
+auto Easy_Client::table_exists() -> bool {
+  // // This does not work for some reason. It returns a table with 0 rows.
+  // auto query = Query{fmt::format("EXISTS TABLE {}", table)};
```

Comment:
Likely because it should be this?

```suggestion
  // This does not work for some reason. It returns a table with 0 rows.
  // auto query = Query{fmt::format("EXISTS TABLE '{}'", table)};
```

(Anyways, what you have below works, so just ignore this.)